### DISCUSSION TITLE
Continue training NER component

### DISCUSSION THREAD

**Discussion by alyserecord** [explosion/spaCy#7955] [Help: Other Questions] [answered]
I am using spaCy 3.0 and the sm english model for a NER problem. I would like to use the sm model and continue to train it using my dataset which contains added custom NER labels and annotations. I am unable to utilize the command line interface for training, so I am having to build my training loop in python. (snippet provided below)

So my questions are:
- is .resume_training() only meant to be used with .rehearse()? I see that the .rehearse has an open issue that I am experiencing (#7161), but I am unclear if calling .update() after .resume_training() is an appropriate path based on the documentation.
- I am apprehensive about using .resume_training() and .rehearse() since they are marked as experimental in v3.0, so is there a different/better way that I could create the optimizer using the trained model and "resume training" (I thought this was already possible with spaCy 2.0 somehow but I cannot find examples now). I understand that the goal of .rehearse() is combat catastrophic forgetting, but I am wondering if there is any other method to continue training the existing model.

```
...
with nlp.disable_pipes(*other_pipes):  
    examples = []
    for text, annots in zip(reviews, annotated_reviews):
        examples.append(Example.from_dict(nlp.make_doc(text), annots))
    optimizer = nlp.resume_training()
    train_examples, test_examples = train_test_split(examples)
    for i in range(n_iter):
        random.shuffle(train_examples)
        for batch in minibatch(train_examples, size=compounding(4., 32., 1.001)):
            nlp.update(batch, sgd=optimizer, drop=dropout)
        print(nlp.evaluate(test_examples))
        print(f"Epoch {i} done.")
```

**Comment by adrianeboyd** [comment, marked as answer]
Hi, for an example of how to continue training the NER component, see the example project [`pipelines/ner_demo_update`](https://github.com/explosion/projects/tree/v3/pipelines/ner_demo_update)

There is an example run-through of a very similar demo project here: https://github.com/explosion/spaCy/tree/master/examples/#-get-started-with-a-demo-project

We'd strongly recommend using the provided training loop over a custom training loop like this because there are lot of good defaults provided in the default configs and training setup.

You don't necessarily have to train through the CLI, you can also call the underlying `train` methods directly if you'd prefer, see:

https://github.com/explosion/spaCy/blob/cf032ec31e38f57940edfb93f041bcd373871554/spacy/cli/train.py#L18-L59

**Reply by alyserecord** [reply]
> We'd strongly recommend using the provided training loop over a custom training loop like this

Are you referring to the train command? Or are you referring to something else I'm not missing from the examples? As previously mentioned I am not able to use the CLI for training like in the examples projects you provided.  


> You don't necessarily have to train through the CLI, you can also call the underlying train methods directly if you'd prefer

Is using train_cli directly the recommended approach (when CLI train is not possible) instead of utilizing the documented methods at `https://spacy.io/api/language`? And if so, is there any documentation on this?

**Reply by adrianeboyd** [reply]
For extending a component in an existing pipeline, this loop is probably fine as long your training data also has all entity types from the original model annotated on the training data. `resume_training` and `update` should be correct. `rehearse` is still definitely experimental and from a glance at the code, most likely currently not working.

A problem that we've seen frequently in v2 is that new users would start with the example training scripts, which were intended more as demos of the basic functionality than for use in a real task, and then kind of get stuck using suboptimal corpus management and training loops because they'd invested some time extending the demo scripts. In addition, the v2 JSON training format used with the v2 train CLI was difficult to work with, so that was a big hurdle to switching to the train CLI.

But in spacy v3, the `DocBin`/`.spacy` training format is much easier to use (see the example `convert.py` scripts in the demo projects) and projects make it much easier to rerun your workflow and reproduce your results in the future. We use projects with CLI commands + a few scripts internally to produce all the pretrained pipelines. The default config settings plus the default training loop manage a lot of details for you, which can make a difference in how well the model trains: batching, shuffling, early stopping, etc.

Just to check in case there's a better solution: can you share why can't you use the CLI? The docs about preferring the train CLI in most cases are here: https://spacy.io/usage/v3#migrating-training. Calling `train_cli` directly would be a bit unusual, but obviously technically no problem.
